### PR TITLE
Add copy to clipboard for preview expression.

### LINKED ISSUE #46160: Add a "copy to clipboard" function for preview values of expression evaluations
### Feature description

The expression dialogs (e. g. the Field Calculator) show a preview of the result value of an entered expression at the bottom. When the stringified value uses many characters, it gets abbreviated after about ~60 characters.

For development/debugging purposes or even just more insight in general, it would be useful to be able to get the full value of the evaluated expression.

For that a "copy to clipboard" icon would be great. It could be conditionally hidden if the value is unabbreviated as the text is selectable and thus copyable already.

If an icon was too much clutter, I propose copying the whole value if the user selects the preview value with doubleclick and hits Ctrl-C (or make the preview text clickable like a button, then copy the value when clicked?). The mouseover popup should include a short instruction for this, like "Doubleclick to copy full value to clipboard".

### Additional context

Some discussion around this in https://lists.osgeo.org/pipermail/qgis-developer/2021-November/064318.html

### PR DESCRIPTION
**Pull request by ismailsunni**
## Description

Adding copy feature preview expression. It can be useful as mentioned in #46160. Suggestions are welcomed :)

Fix #46160.

Latest feature preview:
image

Old (beginning of PR) preview:
Peek 2022-03-08 12-25 - copy preview expression

Funded by Camptocamp https://www.camptocamp.com/

### CODE REVIEW

File: src/gui/qgsexpressionpreviewwidget.cpp [hunk 1]
```diff
@@ -187,3 +194,11 @@ bool QgsExpressionPreviewWidget::evalError() const
 {
   return mEvalError;
 }
+
+void QgsExpressionPreviewWidget::copyFullExpressionValue()
+{
+  QClipboard *clipboard = QApplication::clipboard();
+  const QVariant value = mExpression.evaluate( &mExpressionContext );
+  QgsDebugMsg( QStringLiteral( "set clipboard: %1" ).arg( value.toString() ) );
```
Comment (elpaso) [inline src/gui/qgsexpressionpreviewwidget.cpp:202]: better to use QgsDebugMsgLevel

Comment (ismailsunni) [reply src/gui/qgsexpressionpreviewwidget.cpp]: Thanks, @elpaso, fixed.

File: src/ui/qgsexpressionpreviewbase.ui [hunk 2]
```diff
@@ -87,6 +49,59 @@
...
+       <property name="wordWrap">
+        <bool>true</bool>
+       </property>
+       <property name="textInteractionFlags">
+        <set>Qt::LinksAccessibleByMouse|Qt::TextSelectableByMouse</set>
+       </property>
+      </widget>
+     </item>
+     <item>
+      <widget class="QToolButton" name="mCopyPreviewButton">
```
Comment (nyalldawson) [inline src/ui/qgsexpressionpreviewbase.ui:93]: Can you set the autoraise property on this button? It'll look cleaner in the UI

Comment (ismailsunni) [reply src/ui/qgsexpressionpreviewbase.ui:93]: Done!. Thanks, I didn't know this property before :)

**Comment by 3nids** [comment]
Wouldn't a context menu with "copy entry" be a bit less intrusive in the GUI.
To me, it will be seldom used, and one more click is not that bad.

**Comment by andreasneumann** [comment]
> Wouldn't a context menu with "copy entry" be a bit less intrusive in the GUI.

I agree.

**Comment by ismailsunni** [comment]
Hi @3nids @andreasneumann 
Thanks for the feedback!

Yes, there were two options to implement this feature. A button or a context menu. I chose a copy button since it seems easier to implement. AFAIK, we need to create a custom class (inheriting QLabel) to add a context menu. I tried to "reduce" the code (less maintenance). But if it's much better to have a context menu, I will close this PR and tried again :)

Another option is probably, to hide the copy button until the expression result is too long. In this case, if the expression result is not long, we don't see the button just like before. And if the expression preview is too long, an additional button in the same area seems fine since it's already a lot of text (and show the user that you **can** copy the full value). What do you think?

**Comment by 3nids** [comment]
To me, it depends a bit on the goal of this copy action.
If it is only to see a too long value, to me the tooltip (on mouse hovering) is already enough.
If we really want to copy (it might be a nice addition), the copy action should be accessible all the time. And in this case, I am personally in favor of the context menu. 

I don't think to create a subclass of QWidget/Qlabel to achieve this.
Setting the context menu policy to `Qt::ActionsContextMenu` and simply adding action to the widget should do the trick.

**Comment by ismailsunni** [comment]
Hi @3nids and @andreasneumann 
I have changed it to use the context menu as the suggestion. I can't add the tooltip since it's already used to show the full expression value and used to pass a signal (see: QgsExpressionPreviewWidget::setExpressionToolTip)

image

Thanks!

File: src/gui/qgsexpressionpreviewwidget.h [hunk 3]
```diff
@@ -118,6 +119,7 @@ class GUI_EXPORT QgsExpressionPreviewWidget : public QWidget, private Ui::QgsExp
 
   private:
     void setExpressionToolTip( const QString &toolTip );
+    void copyFullExpressionValue();
```
Comment (3nids) [inline src/gui/qgsexpressionpreviewwidget.h:122]: this should be a private slot

File: src/gui/qgsexpressionpreviewwidget.cpp [hunk 4]
```diff
@@ -13,24 +13,30 @@
...
 
-
-
 QgsExpressionPreviewWidget::QgsExpressionPreviewWidget( QWidget *parent )
   : QWidget( parent )
 {
   setupUi( this );
   mPreviewLabel->clear();
+  mPreviewLabel->setContextMenuPolicy( Qt::ActionsContextMenu );
+  mCopyPreviewAction = new QAction( QgsApplication::getThemeIcon( QStringLiteral( "/mActionEditCopy.svg" ) ), "Copy expression value", this );
```
Comment (3nids) [inline src/gui/qgsexpressionpreviewwidget.cpp:33]: this needs to be translated

**Review by 3nids — COMMENTED** [review]
cool, it looks good!
One more thing, you could add the info that it can copied in the tool tip:
'My suppppppper lonnnnnng value' (right-click to copy)

File: src/ui/qgsexpressionpreviewbase.ui [hunk 5]
```diff
@@ -26,44 +26,6 @@
...
-     </property>
-     <property name="wordWrap">
-      <bool>true</bool>
-     </property>
-     <property name="textInteractionFlags">
-      <set>Qt::LinksAccessibleByMouse|Qt::TextSelectableByMouse</set>
-     </property>
-    </widget>
-   </item>
    <item row="0" column="0" colspan="2">
```
Comment (3nids) [inline src/ui/qgsexpressionpreviewbase.ui:29]: is there any change required in the UI? looks like the colspan is useless, no?

**Comment by ismailsunni** [comment]
Hi @3nids 
Update tool tip:

image
image

Thanks for the suggestion

File: src/ui/qgsexpressionpreviewbase.ui (hunk 5 above)
Comment (ismailsunni) [reply src/ui/qgsexpressionpreviewbase.ui]: yes, removed now.

File: src/gui/qgsexpressionpreviewwidget.cpp [hunk 6]
```diff
@@ -13,24 +13,30 @@
...
 
-
-
 QgsExpressionPreviewWidget::QgsExpressionPreviewWidget( QWidget *parent )
   : QWidget( parent )
 {
   setupUi( this );
   mPreviewLabel->clear();
+  mPreviewLabel->setContextMenuPolicy( Qt::ActionsContextMenu );
+  mCopyPreviewAction = new QAction( QgsApplication::getThemeIcon( QStringLiteral( "/mActionEditCopy.svg" ) ), tr( "Copy expression value" ), this );
```
Comment (nyalldawson) [inline src/gui/qgsexpressionpreviewwidget.cpp:33]:
```suggestion
  mCopyPreviewAction = new QAction( QgsApplication::getThemeIcon( QStringLiteral( "/mActionEditCopy.svg" ) ), tr( "Copy Expression Value" ), this );
```

File: src/gui/qgsexpressionpreviewwidget.cpp [hunk 7]
```diff
@@ -157,7 +166,15 @@ void QgsExpressionPreviewWidget::setExpressionToolTip( const QString &toolTip )
...
-  mPreviewLabel->setToolTip( mToolTip );
+  QString copyToolTip = tr( "right-click to copy" );
+  if ( toolTip.isEmpty() )
+  {
+    mPreviewLabel->setToolTip( copyToolTip );
+  }
+  else
+  {
+    mPreviewLabel->setToolTip( mToolTip + " (" + copyToolTip + ")" );
+  }
```
Comment (nyalldawson) [inline src/gui/qgsexpressionpreviewwidget.cpp:177]:
```suggestion
  if ( toolTip.isEmpty() )
  {
    mPreviewLabel->setToolTip( tr( "Right-click to copy" ) );
  }
  else
  {
    mPreviewLabel->setToolTip( tr("%1 (right-click to copy)" ).arg( mToolTip ) );
  }
```

File: src/gui/qgsexpressionpreviewwidget.h [hunk 8]
```diff
@@ -17,6 +17,7 @@
 #define QGSEXPRESSIONPREVIEWWIDGET_H
 
 #include <QWidget>
+#include <QAction>
```
Comment (3nids) [inline src/gui/qgsexpressionpreviewwidget.h:20]: this can be a forward declaration, no need to include

Comment (ismailsunni) [reply src/gui/qgsexpressionpreviewwidget.h]: Updated.

**Review by nyalldawson — APPROVED** [review]

**Comment by kannes** [comment]
Great stuff, thank you @ismailsunni !

**Comment by pigreco** [comment]
It does not copy ARRAY
2022-03-17_09h21_48
s

**Pull request merged**